ENH: Allow compression in NDFrame.to_csv to be a dict with optional arguments (#26023)

drew-heenan opened the pull request:
- [x] closes #26023 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

drew-heenan committed:
ENH/BUG: Add arcname to to_csv for ZIP compressed csv filename (#26023)

drew-heenan committed:
DOC: Updated docs for arcname in NDFrame.to_csv (#26023)

drew-heenan committed:
conform to line length limit

WillAyd commented:
I am -1 on adding another parameter here. What makes Zip extraction different than the other compressions methods?

drew-heenan commented:
Other compression methods apply to a single file and add an extension to the name (i.e. `data.csv.gz` extracted results in `data.csv`), but ZIP archives are meant to have each file within have a name. As is, calling `df.to_csv('data.zip')` creates an archive containing a CSV file named `data.zip` as well.

drew-heenan committed:
Fixed test_to_csv_zip_arcname for Windows paths

gfyoung commented:
So I understand the use-case.  However, I do agree that adding another parameter is not necessarily the best way to do this.  This parameter is really related to our `compression` parameter.

This leads me to think that a `dict` might make more sense.  It allows us to configure the compression without the bloat.  What do you think?

cc @jreback

drew-heenan committed:
Merge remote-tracking branch 'upstream/master' into issue-26023

drew-heenan commented:
@gfyoung I agree that this approach makes more sense - I'll modify the functions to optionally take a `dict` for `compression`, i.e. `{'compression': 'zip', 'arcname': 'data.csv'}`. Thanks for the feedback!

gfyoung commented:
@drew-heenan : Sounds good!  One minor thing: let's use `method` as the key instead of `compression`.  Otherwise, we'll be extracting `compression["compression"]`, which is a little redundant.

drew-heenan committed:
to_csv compression may now be dict with possible keys 'method' and 'arcname'

gfyoung commented on pandas/io/common.py:
```diff
@@ -329,27 +355,31 @@ def _get_handle(path_or_buf, mode, encoding=None, compression=None,
+        elif compression_method == 'zip':
+            arcname = None
+            if isinstance(compression, dict) and 'arcname' in compression:
+                arcname = compression['arcname']
```
For greater flexibility, pop out `method` and pass in `compression` as `kwargs`.

drew-heenan replied:
Good idea; To clarify, do you mean to add an additional parameter to `ByteZipFile.__init__` which specifies kwargs for `ByteZipFile.write`'s super call to `ZipFile.writestr`? I originally went with passing `arcname` to `ByteZipFile.__init__` as it may be wrapped in a `TextIOWrapper`, which does not support passing kwargs in `write` to the wrapped `ByteZipFile`.  `ZipFile` itself requires that the archived file name be passed in `write`, so I don't think passing it as `kwargs` to `ByteZipFile` would work out.

gfyoung replied:
Something like this:

~~~
compression.pop("method")
zf = BytesZipFile(path_or_buf, mode, **compression)
~~~

gfyoung commented on pandas/io/common.py:
```diff
@@ -299,10 +314,21 @@ def _get_handle(path_or_buf, mode, encoding=None, compression=None,
+        .. versionchanged:: 0.25.0
+
+           May now be a dict with key 'method' as compression mode
+           and 'arcname' as CSV file name if mode is 'zip'
```
I would make this a little more generic.  In reality, we should be just accepting any keyword arguments to `BytesZipFile`.

jreback commented on pandas/io/common.py:
```diff
@@ -368,14 +398,9 @@ def _get_handle(path_or_buf, mode, encoding=None, compression=None,
+        elif compression_method == 'xz':
             f = lzma.LZMAFile(path_or_buf, mode)
 
-        # Unrecognized Compression
```
you still need to catch this case

jreback commented on pandas/io/formats/csvs.py:
```diff
@@ -36,8 +36,20 @@ def __init__(self, obj, path_or_buf=None, sep=",", na_rep='',
         if path_or_buf is None:
             path_or_buf = StringIO()
 
+        self._compression_arg = compression
```
didn't you change get_filepath_or_buffer to already handle this? why is this special cased here?

drew-heenan replied:
I did not modify `get_filepath_or_buffer`, though I suppose I certainly could to support taking a `dict` as `compression`. The `self._compression_arg` is there to avoid changing `self.compression` from only ever holding the inferred compression method, while `self._compression_arg` would include any additional arguments if `compression` was passed as a `dict`. I do now think that I should have instead had a `self.compression_args` hold this dict with the `method` key popped.

drew-heenan replied:
@jreback With regard to `get_filepath_or_buffer` (and similarly `_infer_compression`), I've added a function `_get_compression_method` which handles the case where `compression` is given as a `dict` to `to_csv`. It extracts the compression method string before passing to `get_filepath_buffer` or `_infer_compression`. 
Would it be preferable then to keep both functions' original functionality where `compression` may only be a string/None, or include a call to `_get_compression_method` in both to handle dicts?

jreback commented on pandas/tests/io/formats/test_to_csv.py:
```diff
@@ -537,3 +537,32 @@ def test_to_csv_compression(self, compression_only,
                                  compression=read_compression)
             tm.assert_frame_equal(result, df)
+
+    @pytest.mark.parametrize("method", ["gzip", "bz2", "zip", "xz"])
```
use the fixture for this

WillAyd commented on pandas/io/common.py:
```diff
@@ -329,28 +355,32 @@ def _get_handle(path_or_buf, mode, encoding=None, compression=None,
-        elif compression == 'zip':
-            zf = BytesZipFile(path_or_buf, mode)
+        elif compression_method == 'zip':
+            arcname = None
```
Can you rename this to `archive_name`?

drew-heenan replied:
@WillAyd The specific instance of `arcname` above is no longer there; Though, do you mean change `arcname` to `archive_name` in general, including the dict key?

WillAyd commented on pandas/io/common.py:
```diff
@@ -420,13 +445,16 @@ class BytesZipFile(zipfile.ZipFile, BytesIO):  # type: ignore
     def write(self, data):
-        super(BytesZipFile, self).writestr(self.filename, data)
+        arcname = self.filename if self.arcname is None else self.arcname
+        super(BytesZipFile, self).writestr(arcname, data)
```
We don't need the Py2 compat syntax anymore should just `super()` here is fine

WillAyd reviewed (changes requested):
Could you also add type annotations for the changed / added parameters?

drew-heenan committed:
test_to_csv_compression_dict uses compression_only fixture

drew-heenan committed:
delegate dict handling to _get_compression_method, type annotations

drew-heenan committed:
fix import order, None type annotations

drew-heenan commented:
@WillAyd I've added type annotations to parameters which I've changed, but it seems that doing so caused the typing validation check to fail on other parts of the files which I have not modified.

drew-heenan committed:
compression args passed as kwargs, update relevant docs

WillAyd commented:
Yea I think a general rename would make things clearer

Sent from my iPhone

> On Apr 13, 2019, at 6:24 PM, Drew Heenan <notifications@github.com> wrote:
> 
> @drew-heenan commented on this pull request.
> 
> In pandas/io/common.py:
> 
> >              if is_path:
>                  f = bz2.BZ2File(path_or_buf, mode)
>              else:
>                  f = bz2.BZ2File(path_or_buf)
>  
>          # ZIP Compression
> -        elif compression == 'zip':
> -            zf = BytesZipFile(path_or_buf, mode)
> +        elif compression_method == 'zip':
> +            arcname = None
> @WillAyd The specific instance of arcname above is no longer there; Though, do you mean change arcname to archive_name in general, including the dict key?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

WillAyd commented on pandas/core/generic.py:
```diff
@@ -2918,7 +2919,8 @@ def to_latex(self, buf=None, columns=None, col_space=None, header=True,
                columns=None, header=True, index=True, index_label=None,
-               mode='w', encoding=None, compression='infer', quoting=None,
+               mode='w', encoding=None,
+               compression: Union[str, Dict, None] = 'infer', quoting=None,
```
Should break this up so parameters with annotations exist on their own line

WillAyd commented on pandas/io/common.py:
```diff
@@ -233,13 +234,45 @@ def file_path_to_url(path):
+    ------
+    ValueError on dict missing 'method' key
+    """
+    compression_args = {}  # type: Dict
```
Can you add the key / value types for the Dict

WillAyd commented on pandas/io/common.py:
```diff
@@ -287,8 +318,9 @@ def _infer_compression(filepath_or_buffer, compression):
-def _get_handle(path_or_buf, mode, encoding=None, compression=None,
-                memory_map=False, is_text=True):
+def _get_handle(path_or_buf, mode, encoding=None,
+                compression: Union[str, Dict, None] = None, memory_map=False,
```
Same here for breaking up signature

WillAyd commented on pandas/io/common.py:
```diff
@@ -318,7 +362,7 @@ def _get_handle(path_or_buf, mode, encoding=None, compression=None,
     try:
         from s3fs import S3File
-        need_text_wrapping = (BytesIO, S3File)
+        need_text_wrapping = (BytesIO, S3File)  # type: Tuple
```
Was this throwing a Typing error? Think MyPy should be able to infer here

drew-heenan replied:
@WillAyd Yeah, it was. Couldn’t tell why, but MyPy couldn’t infer when I added types to the function definition. It happened in a couple places.

WillAyd replied:
What error was it giving you?

drew-heenan replied:
It was giving `pandas/io/common.py:367: error: Incompatible types in assignment (expression has type "Tuple[Type[BytesIO]]", variable has type "Tuple[Type[BytesIO], Any]")` only if any of the function's arguments have type annotations.

WillAyd commented on pandas/io/formats/csvs.py:
```diff
@@ -18,26 +19,32 @@
             path_or_buf = StringIO()
 
+        # Extract compression mode as given, if dict
+        compression, self.compression_args \
```
Can you break at the function signature instead? We don't use explicit line continuations like this in code base

drew-heenan committed:
style/doc improvements, change arcname to archive_name

drew-heenan committed:
Merge branch 'master' into issue-26023

WillAyd commented on pandas/core/generic.py:
```diff
@@ -2909,10 +2910,11 @@ def to_latex(self, buf=None, columns=None, col_space=None, header=True,
-               tupleize_cols=None, date_format=None, doublequote=True,
-               escapechar=None, decimal='.'):
+               mode='w', encoding=None,
+               compression: Union[str, Dict, None] = 'infer',
```
Should always have types for Container classes like Dict. I think this should read `Optional[Union[str, Dict[str, str]]]`

drew-heenan replied:
Right; I think it should be `Optional[Union[str, Dict[str, Optional[str]]]]`, though, as both the method and archive name can be `None`.

WillAyd replied:
What is the use case for an optional value in the dict? Isn't the point of accepting one in the first place for it to always have a `method` key with a `str` value associated with it?

drew-heenan replied:
@WillAyd If there's some `compression_method`, including the `Optional` allows `method` to be `None` so we can have `compression` which is passed to `to_csv` be, for example,
```
compression = dict(method=compression_method, **other_args)
```
instead of something more cumbersome like
```
if compression_method is not None:
  compression = dict(method=compression_method, **other_args)
else:
  compression = None
```
which is admittedly not a huge difference, but passing `None` as either `method` or `archive_name` is supported regardless.

WillAyd commented on pandas/io/common.py:
```diff
@@ -232,13 +233,45 @@ def file_path_to_url(path):
 }
 
 
+def _get_compression_method(compression: Union[str, Dict, None]):
```
Same comment on Dict signature

WillAyd commented on pandas/io/common.py:
```diff
@@ -232,13 +233,45 @@ def file_path_to_url(path):
+    ------
+    ValueError on dict missing 'method' key
+    """
+    compression_args = {}  # type: Dict[str, Any]
```
Is this even necessary? Think creation in the conditional should suffice

drew-heenan replied:
@WillAyd I suppose it can be put in an else statement. I'll make that change.

WillAyd commented on pandas/io/common.py:
```diff
@@ -419,13 +463,19 @@ class BytesZipFile(zipfile.ZipFile, BytesIO):  # type: ignore
     # GH 17778
-    def __init__(self, file, mode, compression=zipfile.ZIP_DEFLATED, **kwargs):
+    def __init__(self, file, mode, compression=zipfile.ZIP_DEFLATED,
+                 archive_name: Union[str, zipfile.ZipInfo, None] = None,
```
Where is zipfile.ZipInfo coming from here?

drew-heenan replied:
@WillAyd A `ZipFile.writestr` can take either a `ZipInfo` or a string as the archived file name `zinfo_or_arcname`,  which `archive_name` is passed to, so I figured that type may as well be allowed here.

WillAyd replied:
Right but internally we are only calling with a str right?

drew-heenan replied:
`archive_name` is whatever is passed at `compression['archive_name']`, so it _could_ be a `ZipInfo`

WillAyd replied:
Can you provide a use case for when this would happen? If allowable then should be reflected in the docstring and with a test

drew-heenan replied:
I can tell that file comments and the file date can be set by passing a `ZipInfo`. Not sure if this is an important enough functionality to include - what do you think?

WillAyd replied:
I'm not sure I follow what you are saying but unless this is a feature you think should be documented and supported it should be removed from the annotation

drew-heenan replied:
Got it - I'll remove that type from the annotation.

WillAyd commented on pandas/io/formats/csvs.py:
```diff
@@ -115,7 +122,7 @@ def __init__(self, obj, path_or_buf=None, sep=",", na_rep='',
         if (isinstance(self.data_index, (ABCDatetimeIndex, ABCPeriodIndex)) and
                 date_format is not None):
-            from pandas import Index
+            from pandas import Index  # type: ignore
```
What error was this giving you?

drew-heenan replied:
Giving `pandas/io/formats/csvs.py:125: error: Module 'pandas' has no attribute 'Index'`; It appears to only occur if there are type annotations on any `CSVFormatter` parameters.

WillAyd replied:
Make sure you try again on master - I think a separate PR should have resolved this already

drew-heenan replied:
The same error appears on master if any type annotations are added.

WillAyd replied:
@drew-heenan can you remove this and push as a new commit? Again thought we resolved this in a separate PR so would like to validate its not a mypy versioning thing between your local environment and what we have on CI

drew-heenan replied:
@WillAyd Just did that - the error still appears.

WillAyd replied:
Hmm OK thanks for confirming. @ryankarlos not sure if you have any insight - thought this would be resolved by #26019 

@drew-heenan this isn't a blocker so OK to add back in the ignore I think; can review separate from this

drew-heenan replied:
@WillAyd Got it, thanks for checking!

WillAyd commented on pandas/tests/io/formats/test_to_csv.py:
```diff
@@ -536,3 +536,32 @@ def test_to_csv_compression(self, compression_only,
+    @pytest.mark.parametrize("compression", ["zip", "infer"])
+    @pytest.mark.parametrize("archive_name", [None, "test_to_csv.csv",
+                                              "test_to_csv.zip"])
+    def test_to_csv_zip_arguments(self, compression, archive_name):
```
Will also want to add a test that checks for an error being thrown when you provide compression a dict without `method` as a key

jreback reviewed (changes requested):
i have yet to fully review

drew-heenan commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -36,6 +36,7 @@ Other Enhancements
 - :class:`RangeIndex` has gained :attr:`~RangeIndex.start`, :attr:`~RangeIndex.stop`, and :attr:`~RangeIndex.step` attributes (:issue:`25710`)
 - :class:`datetime.timezone` objects are now supported as arguments to timezone methods and constructors (:issue:`25065`)
 - :meth:`DataFrame.query` and :meth:`DataFrame.eval` now supports quoting column names with backticks to refer to names with spaces (:issue:`6508`)
+- :meth:`NDFrame.to_csv` now supports dicts as ``compression`` argument with key ``'method'`` being the compression method and others as kwargs of ``ByteZipFile`` when the compression method is ``'zip'``. (:issue:`26023`)
```
@jreback To clarify, do you mean that it should say `Series.to_csv` and `DataFrame.to_csv` instead of just `NDFrame.to_csv`?

gfyoung replied:
Yes, that's what @jreback is trying to say

drew-heenan replied:
Got it - I've changed the doc to list both.

drew-heenan committed:
Merge branch 'master' into issue-26023

drew-heenan committed:
add to_csv example, no method test, Optional types, tweaks; update whatsnew

drew-heenan committed:
remove Index import type ignore

drew-heenan committed:
Revert "remove Index import type ignore"

This reverts commit 68a2b4dad39bc60cb1bcf750ec18ae379d33bb29.

drew-heenan committed:
Merge remote-tracking branch 'upstream/master' into issue-26023

jreback commented:
can u merge master

drew-heenan committed:
Merge branch 'master' into issue-26023

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -37,6 +37,7 @@ Other Enhancements
 - :class:`RangeIndex` has gained :attr:`~RangeIndex.start`, :attr:`~RangeIndex.stop`, and :attr:`~RangeIndex.step` attributes (:issue:`25710`)
 - :class:`datetime.timezone` objects are now supported as arguments to timezone methods and constructors (:issue:`25065`)
 - :meth:`DataFrame.query` and :meth:`DataFrame.eval` now supports quoting column names with backticks to refer to names with spaces (:issue:`6508`)
+- :meth:`DataFrame.to_csv` and :meth:`Series.to_csv` now support dicts as ``compression`` argument with key ``'method'`` being the compression method and others as kwargs of ``ByteZipFile`` when the compression method is ``'zip'``. (:issue:`26023`)
```
what is a ByteZipFile (meaning a user has NO idea what this is); pls rephrase.

jreback commented on pandas/core/generic.py:
```diff
@@ -2961,17 +2962,21 @@ def to_csv(self, path_or_buf=None, sep=",", na_rep='', float_format=None,
+            detect compression mode from the following extensions: '.gz',
+            '.bz2', '.zip' or '.xz'. (otherwise no compression). If dict given
+            and mode is 'zip' or inferred as 'zip', other entries passed as
+            kwargs to ByteZipFile.
```
ByteZipFile is not user facing, see my comments above

jreback commented on pandas/core/generic.py:
```diff
@@ -3023,6 +3028,10 @@ def to_csv(self, path_or_buf=None, sep=",", na_rep='', float_format=None,
         ...                    'weapon': ['sai', 'bo staff']})
         >>> df.to_csv(index=False)
         'name,mask,weapon\nRaphael,red,sai\nDonatello,purple,bo staff\n'
+        >>> compression_opts = dict(method='zip', archive_name='out.csv')
```
can you put the comment before the example (and put a blank line between cases); also might need to have a DOCTEST: SKIP here

jreback replied:
can you do this

jreback reviewed (changes requested):
looks reasonable, some comments

drew-heenan committed:
Improve docs/examples

drew-heenan committed:
Merge branch 'master' into issue-26023

WillAyd commented on pandas/io/common.py:
```diff
@@ -232,13 +233,46 @@ def file_path_to_url(path):
 }
 
 
+def _get_compression_method(compression: Optional[Union[str, Dict[str, Any]]]):
```
I think the Dict can be typed as `Dict[str, str]` here. Would also be nice to type the return though that may change depending on comment below

jreback replied:
can you do this

WillAyd commented on pandas/io/common.py:
```diff
@@ -232,13 +233,46 @@ def file_path_to_url(path):
+                             "must have key 'method'")
+    else:
+        compression_args = {}
+    return compression, compression_args
```
Any reason to return compression and compression_args here? The former will always match whatever is being passed to the function anyway so seems like we could move the `isinstance` check outside of this function and only call this method when `compression` is a dict

WillAyd commented on pandas/io/common.py:
```diff
@@ -328,6 +372,7 @@ def _get_handle(path_or_buf, mode, encoding=None, compression=None,
     path_or_buf = _stringify_path(path_or_buf)
     is_path = isinstance(path_or_buf, str)
 
+    compression, compression_args = _get_compression_method(compression)
```
Related to comment above, I think it would make sense to do the isinstance check before calling this method, unless you think there is a reason why we would ever need this method to do anything with a `str` passed through as an argument

WillAyd commented on pandas/tests/io/formats/test_to_csv.py:
```diff
@@ -534,3 +534,39 @@ def test_to_csv_compression(self, compression_only,
+            read_df = pd.read_csv(path, index_col=0)
+            tm.assert_frame_equal(read_df, df)
+
+    def test_to_csv_compression_dict_no_method(self):
```
Can you append `_raises` to this test name?

WillAyd commented on pandas/tests/io/formats/test_to_csv.py:
```diff
@@ -534,3 +534,39 @@ def test_to_csv_compression(self, compression_only,
+        # GH 26023
+        df = DataFrame({"ABC": [1]})
+        compression = {"some_option": True}
+        with tm.ensure_clean("out.zip") as path, pytest.raises(ValueError):
```
Can you match on the expected message with `pytest.raises` here?

jreback replied:
do the raises as an inner context manager,e .g.
```
with tm.ensure_clearn(...):
     with pytest.raises(...., msg=msg):
```

WillAyd commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -38,6 +38,7 @@ Other Enhancements
 - :class:`RangeIndex` has gained :attr:`~RangeIndex.start`, :attr:`~RangeIndex.stop`, and :attr:`~RangeIndex.step` attributes (:issue:`25710`)
 - :class:`datetime.timezone` objects are now supported as arguments to timezone methods and constructors (:issue:`25065`)
 - :meth:`DataFrame.query` and :meth:`DataFrame.eval` now supports quoting column names with backticks to refer to names with spaces (:issue:`6508`)
+- :meth:`DataFrame.to_csv` and :meth:`Series.to_csv` now support dicts as ``compression`` argument with key ``'method'`` being the compression method and others as additional compression options when the compression method is ``'zip'``. (:issue:`26023`)
```
It's hinted at but not explicitly stated that the `dict` only works when the `method` key is `zip` right? If so can we clarify that via the docstrings and make sure we raise if another method is supplied? I think current implementation just swallows other kwargs for other compression methods

jreback commented:
can you merge master and update to comments

jreback commented:
@drew-heenan can you update

@pandas-dev/pandas-core if someone can merge master

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into issue-26023

WillAyd commented:
Merged master to keep updated. @drew-heenan let us know if can continue with comments!

WillAyd committed:
Added back missed Callable import in generic

jreback commented on pandas/io/common.py:
```diff
@@ -219,13 +220,46 @@ def file_path_to_url(path):
+
+    Returns
+    -------
+    tuple of ({compression method}, any
```
any -> Optional[str]
dict -> Dict[str, str]

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into issue-26023

WillAyd committed:
Address comments

WillAyd committed:
Typing cleanup

WillAyd committed:
Cleaned up docstring

WillAyd commented:
@jreback comments addressed

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into issue-26023

TomAugspurger commented:
@jreback @WillAyd do you recall how close this was to being finished? Probably just need to fix merge conflicts from black?

gfyoung commented:
@TomAugspurger : This was pretty close (especially since the comments later reduced to those regarding annotations).  I think if we fix merge conflicts and blacken this, we should be good.

WillAyd commented:
I can fix up today

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into issue-26023

pep8speaks commented:
Hello @drew-heenan! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/io/common.py`](https://github.com/pandas-dev/pandas/blob/7717f1648af4c62a7d48117ae9a9965949c7e882/pandas/io/common.py):

> [Line 333:89](https://github.com/pandas-dev/pandas/blob/7717f1648af4c62a7d48117ae9a9965949c7e882/pandas/io/common.py#L333): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (126 > 88 characters)

WillAyd committed:
blackify

gfyoung commented:
@WillAyd : Do you plan on annotating on affected methods in this PR?  I see that some of the inputs to some of the affected methods are annotated, but not all of them.

WillAyd commented:
I can take a look

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into issue-26023

WillAyd committed:
Added annotations where feasible

WillAyd committed:
Black and lint

WillAyd commented:
OK @gfyoung added annotations where I think feasible. Two blockers prevented full annotations of modified funcs, particularly _get_handle:

  1. An actual bug in type shed (see https://github.com/python/typeshed/issues/3124)
  2. Variable reuse (see `f` in `_get_handle`); this requires a refactor that I think expands the diff a little much so better done as follow up

WillAyd commented on pandas/io/common.py:
```diff
@@ -435,13 +486,23 @@ class BytesZipFile(zipfile.ZipFile, BytesIO):  # type: ignore
 
     # GH 17778
-    def __init__(self, file, mode, compression=zipfile.ZIP_DEFLATED, **kwargs):
+    def __init__(
```
It might not be clear in the diff but note that I removed the `compression` argument here. The reason for this was that it is never actually called in code. 

It makes annotations more complex, because the keyword argument unpacking of `compression_args` in this PR only ever has `str` as keys and therefore mypy complains that there is a type mismatch because `int` values are never unpacked. Figured easier to just remove than muck around types since it is not ever used

gfyoung commented on pandas/io/common.py:
```diff
@@ -242,13 +244,46 @@ def file_path_to_url(path):
     the inferred compression method is returned. Otherwise, the input
     compression method is returned unchanged, unless it's invalid, in which
     case an error is raised.
-
```
Why are these lines being deleted?

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into issue-26023

WillAyd committed:
isort fixup

WillAyd committed:
Docstring fixup and more annotations

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into issue-26023

WillAyd commented:
Merged again to keep fresh. There's a Mypy failure I'll have to look at later.

@TomAugspurger @gfyoung would you still have time to review this one coming up? Have a lot in the queue so just want to prioritize workflow with you

gfyoung commented:
It looks pretty good overall!

WillAyd committed:
lint fixup

WillAyd committed:
mypy fixup

WillAyd committed:
whatsnew fixup

WillAyd committed:
Annotation and doc fixups

WillAyd committed:
mypy typeshed bug fix

WillAyd commented on pandas/io/common.py:
```diff
@@ -312,32 +357,49 @@ def _infer_compression(
-    path_or_buf, mode, encoding=None, compression=None, memory_map=False, is_text=True
+    path_or_buf,
+    mode: str,
+    encoding=None,
```
Couldn't annotate this particular argument due to a minor bug in typeshed. Fixed on master so maybe something we can come back to soon (typeshed updates are pretty quick)

see https://github.com/python/typeshed/pull/3125

WillAyd reviewed (commented):
OK I think good for review again @gfyoung
